File: Problems/Arrays/Two-Sum/Two-Sum.cpp
#include <iostream>
#include <vector>
#include <map>

// Using hash table - Time: O(n)

class Solution {
public:
    std::vector<int> twoSum(std::vector<int> &nums, int target)
    {
        std::map<int, int> map;
        
		for (int i = 0; i < nums.size(); i++)
        {
            int t = target - nums[i];

            // If the map contains the key t, return the index of t and the current index i

            if (map.count(t))
                return {map[t], i}; 

            map[nums[i]] = i;       
        }

        return {}; 
    }
};

int main()
{
    std::vector<int> nums = {-1, 0, 1, 2, -1, -4};
    int target = 0;

    std::cout << "Input: nums = [";
    for (const auto &num : nums)
    {
        std::cout << num << "";
        if (&num != &nums.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "], target = " << target << std::endl;

    Solution sol;
    std::vector<int> result = sol.twoSum(nums, target);

    std::cout << "Output: [";
    for (const auto &r : result)
    {
        std::cout << r << "";
        if (&r != &result.back())
        {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;

    return 0;
}

File: Problems/Arrays/Two-Sum/Two-Sum.cs
using System;
using System.Collections.Generic;

// Using hash table - Time: O(n)

public class Solution /* Define a public class Solution */
{
	public int[] TwoSum(int[] nums, int target)
	{
		var dic = new Dictionary<int, int>();

		for (int i = 0; i < nums.Length; i++)
		{
			int t = target - nums[i];

			// If the dictionary contains the key t, return the index of t and the current index i

			if (dic.ContainsKey(t)) 
				return new int[] { dic[t], i };

			dic[nums[i]] = i;
		}

		return new int[] { };
	}
}

class Program
{
    static void Main(String[] args)
    {
        int[] nums = { 2, 7, 11, 15 };
        int target = 9;

        Console.Write("Input: nums = [");
        foreach (int n in nums)
        {
            Console.Write(n + "");
            if (n != nums[nums.Length - 1])
            {
                Console.Write(", ");
            }
        }
        Console.WriteLine("], target = " + target);

        Solution sol = new Solution();
        int[] result = sol.TwoSum(nums, target);

        Console.Write("Output: [");
        foreach (int r in result)
        {
            Console.Write(r + "");
            if (r != result[result.Length - 1])
            {
                Console.Write(", ");
            }
        }
		Console.WriteLine("]");
    }
}





